osquery/tables/networking/windows/windows_firewall_rules.cpp

```diff
@@ -0,0 +1,411 @@
+    if (FAILED(
+            hr = getString(rule, &INetFwRule::get_LocalPorts, r.localPorts))) {
+      return createWindowsFirewallRuleError(
+          WindowsFirewallError::RuleLocalPortsError, hr);
+    }
+
+    if (FAILED(hr = getString(
+                   rule, &INetFwRule::get_RemotePorts, r.remotePorts))) {
+      return createWindowsFirewallRuleError(
+          WindowsFirewallError::RuleRemotePortsError, hr);
+    }
+
+  } else {
+    if (FAILED(hr = getString(rule,
+                              &INetFwRule::get_IcmpTypesAndCodes,
+                              r.icmpTypesCodes))) {
+      return createWindowsFirewallRuleError(
+          WindowsFirewallError::RuleICMPTypesCodesError, hr);
+    }
+  }
+
+  if (FAILED(hr = rule->get_Profiles(&r.profileBitmask))) {
+    return createWindowsFirewallRuleError(
+        WindowsFirewallError::RuleProfilesError, hr);
+  }
+
+  return r;
+}
+
+Expected<WindowsFirewallRule, WindowsFirewallError> getNext(
+    IEnumVARIANT* enumvar, bool& found) {
+  found = false;
+
+  WindowsFirewallRule winFwRule;
+  Expected<WindowsFirewallRule, WindowsFirewallError> r(winFwRule);
+
+  ULONG fetched = 0;
+
+  VARIANT value;
+  VariantInit(&value);
+
+  INetFwRule* rule = NULL;
+
+  HRESULT hr = enumvar->Next(1, &value, &fetched);
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRuleError(
+        WindowsFirewallError::PolicyRulesEnumNextError, hr);
+  } else {
+    if (hr != S_FALSE) {
+      found = true;
+      hr = VariantChangeType(&value, &value, 0, VT_DISPATCH);
+      if (FAILED(hr)) {
+        // Log only
+        auto err = createWindowsFirewallRuleError(
+            WindowsFirewallError::PolicyRulesEnumIDispatchError, hr);
+        TLOG << err.getError().getMessage();
+      } else {
+        hr = (V_DISPATCH(&value))->QueryInterface(IID_PPV_ARGS(&rule));
+        if (FAILED(hr)) {
+          // Log only
+          auto err = createWindowsFirewallRuleError(
+              WindowsFirewallError::PolicyRuleInterfaceError, hr);
+          TLOG << err.getError().getMessage();
+        } else {
+          r = populateFirewallRule(rule);
+        }
+      }
+    }
+  }
+
+  // Cleanup
+  if (rule != NULL) {
+    rule->Release();
+  }
+
+  VariantClear(&value);
+
+  return r;
+}
+
+Expected<WindowsFirewallRules, WindowsFirewallError> getFirewallRules(
+    QueryContext& context) {
+  WindowsFirewallRules results;
+  Expected<WindowsFirewallRules, WindowsFirewallError> r(results);
+
+  INetFwPolicy2* netFwPolicy = NULL;
+  INetFwRules* rules = NULL;
+  IUnknown* enumerator = NULL;
+  IEnumVARIANT* enumvar = NULL;
+
+  HRESULT hr = CoCreateInstance(__uuidof(NetFwPolicy2),
+                                NULL,
+                                CLSCTX_INPROC_SERVER,
+                                IID_PPV_ARGS(&netFwPolicy));
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRulesError(WindowsFirewallError::NetFwPolicyError,
+                                        hr);
+    goto Cleanup;
```

Comment:
`goto` statements should not be used (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-goto).